src/threading/MsgThread.cc

```diff
@@ -165,6 +165,72 @@ bool ReporterMessage::Process() {
     return true;
 }
 
+
+// This is the IO source used by MsgThread.
+//
+// The lifetime of the IO source is decoupled from
+// the thread. The thread may be terminated prior
+// to the IO source being properly unregistered and
+// forgotten by the IO manager. Specifically the
+// threading manager would delete an IO source which
+// the IO manager still believed to be ready.
+//
+// See issue #3682 for more details.
+class MsgThread_IOSource : public iosource::IOSource {
```

Comment:
The name feels a bit clunky to me. Could move the class into `MsgThread` so that it becomes `MsgThread::IOSource`, or into namespace, e.g., `threading::IOSource`?